crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -1,14 +1,404 @@
-use std::marker::PhantomData;
+use {
+    crate::{
+        component::{
+            func::{self, LiftContext, LowerContext},
+            matching::InstanceType,
+            values::{ErrorContextAny, FutureAny, StreamAny},
+            Val,
+        },
+        vm::component::{
+            ComponentInstance, LocalErrorContextRefCount, StateTable, StreamFutureState,
+            WaitableState,
+        },
+        AsContextMut,
+    },
+    anyhow::{bail, Result},
+    std::{marker::PhantomData, mem::MaybeUninit},
+    wasmtime_environ::component::{
+        CanonicalAbiInfo, InterfaceType, TypeFutureTableIndex, TypeStreamTableIndex,
+    },
+};
+
+#[derive(Copy, Clone, Debug)]
+enum TableIndex {
+    Stream(TypeStreamTableIndex),
+    Future(TypeFutureTableIndex),
+}
+
+fn state_table(instance: &mut ComponentInstance, ty: TableIndex) -> &mut StateTable<WaitableState> {
+    let runtime_instance = match ty {
+        TableIndex::Stream(ty) => instance.component_types()[ty].instance,
+        TableIndex::Future(ty) => instance.component_types()[ty].instance,
+    };
+    &mut instance.component_waitable_tables()[runtime_instance]
+}
+
+fn get_mut_by_index_from(
+    state_table: &mut StateTable<WaitableState>,
+    ty: TableIndex,
+    index: u32,
+) -> Result<(u32, &mut StreamFutureState)> {
+    Ok(match ty {
+        TableIndex::Stream(ty) => {
+            let (rep, WaitableState::Stream(actual_ty, state)) =
+                state_table.get_mut_by_index(index)?
+            else {
+                bail!("invalid stream handle");
+            };
+            if *actual_ty != ty {
+                bail!("invalid stream handle");
+            }
+            (rep, state)
+        }
+        TableIndex::Future(ty) => {
+            let (rep, WaitableState::Future(actual_ty, state)) =
+                state_table.get_mut_by_index(index)?
+            else {
+                bail!("invalid future handle");
+            };
+            if *actual_ty != ty {
+                bail!("invalid future handle");
+            }
+            (rep, state)
+        }
+    })
+}
+
+fn host_close_reader<U, S: AsContextMut<Data = U>>(store: S, rep: u32) -> Result<()> {
+    _ = (store, rep);
+    todo!()
+}
 
 /// Represents the readable end of a Component Model `future`.
 pub struct FutureReader<T> {
+    rep: u32,
```

Comment:
Mind adding some comments as to what this `rep` is? My naive understanding of component model async is probably showing here but I'm familiar with "rep" of a resource where it's user-supplied but I didn't think that there was anything user-supplied here but rather this is an index to something (but I'm not sure)

Reply:
How about something like `rep: WaitableIndex`? Basically using a type here to indicate more clearly what it's pointing to. My history with resources makes me think that "rep" comes from guests themselves and needs to be preserved and handed back to the guest but I don't think that's what's going on here, so this isn't so much a "representation" internally but moreso an index somewhere else, which is where I think a custom newtype might help to distinguish